### ISSUE TITLE
Misreading of PP files

### ISSUE THREAD

**Issue by dgaleareading** [NCAS-CMS/cf-python#161]
Hi,

I am trying to read in a .pp file outputted from a UM run but two variables, U and V, had some missing values. I think that cf is not handling the file correctly. After converting the PP file to an NC file, no values were missing. Below is a plot of U at 850hPa comparing the two results:

![compare](https://user-images.githubusercontent.com/45002049/101021293-a3be6680-3567-11eb-820d-028d2b5400ea.png)

The code used to generate the above plot is the following:

```
import cf
import numpy as np
import matplotlib.pyplot as plt

nc = cf.read("bz203a.pc19890101.pp")

pp_field = nc.select("eastward_wind")[-4]
pp_field.dump()

nc = cf.read("bz203a.pc19890101.nc")

nc_field = nc.select("eastward_wind")[1]
nc_field.dump()

plt.figure(figsize=(15, 5))

plt.subplot(121)
plt.title("PP File")
plt.contourf(pp_field.data.array[0, :, :])
plt.colorbar()

plt.subplot(122)
plt.title("NC File")
plt.contourf(nc_field.data.array[0, 0, :, :])
plt.colorbar()

plt.savefig("compare.png")
```

The example files are on JASMIN at `/work/scratch-nopw/dg/example` along with the code to create the example and the script used to convert the PP file to an NC file.

Environment (via `cf.environment(paths=False)`):

Platform: Linux-3.10.0-1127.19.1.el7.x86_64-x86_64-with-centos-7.8.2003-Core
HDF5 library: 1.10.4
netcdf library: 4.6.3
udunits2 library: /opt/jaspy/lib/libudunits2.so.0
python: 3.7.1
netCDF4: 1.5.3
cftime: 1.1.1.2
numpy: 1.17.0
psutil: 5.6.3
scipy: 1.4.1
matplotlib: 3.1.1
ESMF: not available
cfdm: 1.8.0
cfunits: 3.2.5
cfplot: not available
cf: 3.2.0

**Comment by sadielbartholomew** [comment]
Hi @dgaleareading, thanks for reporting this to us and providing a lot of detail to allow us to easily understand the issue and go on to investigate.

This does indeed look like a bug. I'll grab that file from JASMIN and see if I can recreate the output described. If so, I'll investigate a fix.

Out of interest, is there a reason you are using matplotlib to plot the fields, rather than [cfplot](https://ajheaps.github.io/cf-plot/) which in many circumstances, such as this I believe via [the `cfplot.con` method](http://ajheaps.github.io/cf-plot/user_guide.html#contour-plots), would allow you to achieve the same plot without having to do a lot of the formatting by hand? (Perhaps it is simply to illustrate that this is a `cf` bug if at all rather than anything related to cfplot?)

**Comment by dgaleareading** [comment]
Hi @sadielbartholomew,

thanks for having a look at this.

II am aware of cf-plot but was doing some processing on the underlying numpy array (via `field.data.array`) so I used matplotlib to make sure I was getting the results I expected.

Thanks,
Daniel

**Comment by sadielbartholomew** [comment]
@dgaleareading Ah I see, that's all good. Sadly I haven't been able to start looking into this yet as I was having trouble transferring the file from `scratch` on JASMIN to my local machine but have managed to grab it now so will have a look shortly.

**Comment by davidhassell** [comment]
Hi @dgaleareading I notice that you're using a old version of cf-python - it might be worth trying with the latest (3.7.0) - it might work now (!?), but if it still doesn't, any bug fix will be built on 3.7.0 for release in 3.8.0.

Thanks,
David

**Comment by dgaleareading** [comment]
Hi @davidhassell, I updated my cf-python and `cf.environment(paths=False)` shows this output:

Platform: Linux-3.10.0-1127.19.1.el7.x86_64-x86_64-with-centos-7.8.2003-Core
HDF5 library: 1.10.4
netcdf library: 4.6.3
udunits2 library: /apps/jasmin/jaspy/miniconda_envs/jaspy3.7/m3-4.6.14/envs/jaspy3.7-m3-4.6.14-r20200606/lib/libudunits2.so.0
python: 3.7.6
netCDF4: 1.5.3
cftime: 1.3.0
numpy: 1.17.0
psutil: 5.6.3
scipy: 1.4.1
matplotlib: 3.1.1
ESMF: 8.0.0
cfdm: 1.8.7.0
cfunits: 3.3.1
cfplot: 3.0.20
cf: 3.7.0

however the problem still persists.

Thanks,
Daniel

**Comment by sadielbartholomew** [comment]
Good spot @davidhassell, I missed that. Thanks for letting us know the issue persists even on the latest version Daniel, we will get back to you once we've investigated.

**Comment by sadielbartholomew** [comment]
Hi @dgaleareading, just to get back to you with my findings. In short, after investigation I'm confident this is *not* a bug; cf-python is in fact handling the file correctly but the missing data you are seeing is inherent in the individual PP fields of the dataset.

I've tried to justify this all, and in more detail, below. I've also sought a second opinion from @davidhassell and he agrees given the reasoning below that it is not a bug. I hope you will agree based on the below (saying that, it is completely understandable that you thought it was a bug).

Therefore, I'm going to remove the 'bug' label shortly and close this Issue unless you can provide a reason as to why there is behaviour that could be consider buggy.


### Justification for it not being a bug

The reason your plot above is showing different results is that the right-hand netCDF output is the result of many of the PP files aggregated, whereas the left-hand PP output is from a single field. The conversion from PP to netCDF was obscuring this fact. The individual PP fields do have missing data resulting in those "stripes" appearing on the plots.

To prove that the PP file is being processed and displayed properly, I tweaked your script to convert form PP to netCDF using cf-python itself, notably:

* specifying `aggregate=False` so the conversion wouldn't combine many of the fields as in your own conversion;
* sorting the fields in the FieldList before using them so equivalent fields map by index for the read-in PP and converted netCDF (I had to use a data point as a sort key since there was no metadata property that appeared to be unique for each field in the list to use to ensure the same order in PP and netCDF) ;
* using the `cf.read( ... ,select="eastward_wind", ...)` equivalent to your  `<Field>.select("eastward_wind")` use (see note on deprecation below);
* throwing the plotting part of your script into a function that accepts an index on the FieldList to easily plot to compare given comparable fields.

From that, I saw that the data as plotted is identical for each of those corresponding fields drawn from the read-in PP and from the converted netCDF field, including many fields of the netCDF showing the same stripes pattern of missing data as with the PP. As an example to illustrate, I've included the outputs for fields at indices 5, 15 and 20 below. Hopefully that is enough to convince you, but feel free to copy the script I've included and try out other indices.

As a side note, FYI, the `select` method on a FieldList has been deprecated since version 3.0.0, being replaced with a `select` keyword argument to the `read` method (as documented under 'Parameters' in the API reference [here](https://ncas-cms.github.io/cf-python/function/cf.read.html?highlight=read#cf.read)). (I'm not sure why the `select` method worked in this particular case, but it shouldn't have and doesn't work for me elsewhere :thinking: .)


### Outputs from script as given below

#### Fields at index 5:

![compare-field-at-index-5](https://user-images.githubusercontent.com/30274190/101453899-fa40f180-3927-11eb-853c-1e0b05f73c65.png)


#### ...at index 15:

![compare-field-at-index-15](https://user-images.githubusercontent.com/30274190/101453906-fca34b80-3927-11eb-8605-c12cadc4a5de.png)


#### ...at index 20:

![compare-field-at-index-20](https://user-images.githubusercontent.com/30274190/101453920-ff05a580-3927-11eb-893c-1f005eba4a5c.png)


### Full working script, amended from your own

```python
import cf
import numpy as np
import matplotlib.pyplot as plt

# Confirm cf-python version
print(cf.__version__)


pp = cf.read("bz203a.pc19890101.pp", select="eastward_wind")
# ... and get netCDF equivalent by converting using cf-python
conversion_filename = "converted_pp_bz203a.nc"
cf.write(pp, conversion_filename)
nc = cf.read(
    conversion_filename, select="eastward_wind", aggregate=False)


# First must sort to get fields in same order, else indices won't correspond
pp.sort(lambda f: f.datum(0))
nc.sort(lambda f: f.datum(0))


def make_comparison_plot(field_index):
    # Take same index on both FieldLists to get corresponding fields:
    pp_field = pp[field_index]
    nc_field = nc[field_index]

    # Plotting to compare and contrast
    plt.figure(figsize=(15, 5))

    plt.subplot(121)
    plt.title("PP File at index {}".format(field_index))
    plt.contourf(pp_field.data.array[0, :, :])  # orig is 0
    plt.colorbar()

    plt.subplot(122)
    plt.title("NC File at index {}".format(field_index))
    plt.contourf(nc_field.data.array[0, :, :])
    plt.colorbar()

    plt.savefig("compare-field-at-index-{}.png".format(field_index))


make_comparison_plot(20)  # identical, both have "stripes"
make_comparison_plot(15)  # ditto
make_comparison_plot(5)  # identical, nethier has "stripes"

plt.show()
```

**Comment by davidhassell** [comment]
Thanks for the nice analysis, @sadielbartholomew.

I think that `cf.FieldLis.select` is not currently deprecated, but is an alias of `cf.FieldList.select_by_identity` (https://github.com/NCAS-CMS/cf-python/blob/v3.7.0/cf/fieldlist.py#L1276) - I think it was once deprecated in the near past, but had a reprieve at some stage!

**Comment by sadielbartholomew** [comment]
Aha, thanks David, that explains why it worked then I was quite puzzled that it did, but I've now realised it's because I had accidentally looked at the docs for that method on `Field` rather than `FieldList` and it does [state it is deprecated](https://ncas-cms.github.io/cf-python/method/cf.Field.select.html?highlight=select#cf.Field.select) for the former.

In that case, it should be fine to use, my bad. It doesn't change the analysis above.

**Comment by dgaleareading** [comment]
Thanks for looking into this @sadielbartholomew @davidhassell. I understand your reasoning and have managed to recreate the examples as shown. 

However, I have plotted the data from the pp file on xconv on ARCHER and all the fields seem to be filled. I have included some of the plots at [xconv_plots.zip](https://github.com/NCAS-CMS/cf-python/files/5672782/xconv_plots.zip). I am not sure if this shows that xconv aggregates the data too for it to show fully filled fields, or if something else is going on.

**Comment by sadielbartholomew** [comment]
Thanks @dgaleareading. Fair enough, it is I suppose an unfair experiment to look at only results from cf-python for that dataset! We are investigating further with other tools.

**Comment by davidhassell** [comment]
Hi - I have found a line in the WGDOS unpacking C code that is different to the equivalent line in other versions of the unpacking - and when I replace it with the what we have elsewhere, cf-python gives the "right" answer (no full rows of zeros, same data values as `xconv`). So, I'm fairly confident that this is the solution, but I'm checking on the what other consequences changing this line will have.

**Comment by davidhassell** [comment]
Hi - I have confirmed that the fix works (see below). This will be released in 3.8.0, due later today ....

The bug only appears when a row has masked elements (thanks to @SimonWNCAS for highlighting this) in it, which explains the geographically sporadic nature of the symptoms in this case.

```python
import cf
import numpy as np
import matplotlib.pyplot as plt

assert(cf.__version__ == '3.8.0')

nc = cf.read("bz203a.pc19890101.pp")

pp_field = nc.select("eastward_wind")[-4]

plt.figure(figsize=(15, 5))

plt.subplot(121)
plt.title("PP File")
plt.contourf(pp_field.data.array[0, :, :])
plt.colorbar()
```
![compare](https://user-images.githubusercontent.com/8126576/102467113-d37d6c00-4047-11eb-9475-fcf256c69916.png)

**Comment by dgaleareading** [comment]
Hi, thanks for figuring this out @davidhassell @SimonWNCAS @sadielbartholomew. Look forward to using the fix in the next update.
